prow/git/v2/client_factory.go

```diff
@@ -339,21 +347,70 @@ func (c *clientFactory) ClientForWithRepoOpts(org, repo string, repoOpts RepoOpt
 		// something unexpected happened
 		return nil, err
 	} else {
-		// we have cloned the repo previously, but will refresh it
-		if err := cacheClientCacher.RemoteUpdate(); err != nil {
-			return nil, err
+		// We have cloned the repo previously, but will refresh it. By default
+		// we refresh all refs with a call to `git remote update`.
+		if repoOpts.FetchCommits == nil {
+			if err := cacheClientCacher.RemoteUpdate(); err != nil {
+				return nil, err
+			}
+		} else {
+			needTargetedFetch = true
 		}
 	}
 
-	// initialize the new derivative repo from the cache
+	// Clone the secondary clone (local clone with checked out files) from the
+	// primary (bare repo).
 	if err := repoClientCloner.CloneWithRepoOpts(cacheDir, repoOpts); err != nil {
 		return nil, err
 	}
 
+	// Here we do a targeted fetch, if the repoOpts for the secondary clone
+	// asked for it.
+	//
+	// TODO(listx): Change ensureCommits() so that it takes a "cacher" interface
+	// instead of a generic RepoClient. However, if ShareObjectsWithSourceRepo
+	// was not set, then we cannot run ensureCommits() with the cacher because
+	// the cacher will run on the primary, not secondary (and without --shared,
+	// this will have no effect on the secondary); we have to handle that case
+	// gracefully (perhaps with some logging also).
+	if needTargetedFetch {
```

Comment:
We can combine these 2 if statements.

Reply:
That would not be the best thing in this situation because it would obscure the importance of the `needTargetedFetch` variable.

Although, now that I'm looking at this code again, I think this variable is not necessary because we can just reuse repoOpts.FetchCommits. Will update.

Reply:
Done.